packages/dds/tree/src/feature-libraries/chunked-forest/codec/codecs.ts

```diff
@@ -12,21 +12,97 @@ import {
 	type IJsonCodec,
 	makeVersionedValidatedCodec,
 } from "../../../codec/index.js";
-import { CursorLocationType, type SchemaAndPolicy } from "../../../core/index.js";
+import {
+	CursorLocationType,
+	type FieldKey,
+	type ITreeCursorSynchronous,
+	type SchemaAndPolicy,
+	type TreeChunk,
+	type TreeNodeSchemaIdentifier,
+} from "../../../core/index.js";
 import type { JsonCompatibleReadOnly } from "../../../util/index.js";
-import { TreeCompressionStrategy } from "../../treeCompressionUtils.js";
+import {
+	TreeCompressionStrategy,
+	TreeCompressionStrategyExtended,
+	type TreeCompressionStrategyInternal,
+} from "../../treeCompressionUtils.js";
 
 import { decode } from "./chunkDecoding.js";
 import type { FieldBatch } from "./fieldBatch.js";
 import { EncodedFieldBatch, validVersions } from "./format.js";
 import { schemaCompressedEncode } from "./schemaBasedEncode.js";
 import { uncompressedEncode } from "./uncompressedEncode.js";
+import { Type, type Static } from "@sinclair/typebox";
+
+/**
+ * Reference ID for a chunk that is incrementally encoded.
+ */
+export const ChunkReferenceId = Type.Number({ multipleOf: 1, minimum: 0 });
+export type ChunkReferenceId = Static<typeof ChunkReferenceId>;
+
+/**
+ * Properties for incremental encoding.
+ * Fields that support incremental encoding will encode their chunks separately by calling `encodeIncrementalField`.
+ * This supports features like incremental summarization where the summary from these fields can be re-used if
+ * unchanged between summaries.
+ * Note that each of these chunks that are incrementally encoded is fully self-describing (contain its own shapes
+ * list and identifier table) and does not rely on context from its parent.
+ * @internal
+ */
+export interface IncrementalEncoder {
+	/**
+	 * Returns whether a field should be incrementally encoded.
+	 * @param nodeIdentifier - The identifier of the node containing the field.
+	 * @param fieldKey - The key of the field to check.
+	 */
+	shouldEncodeFieldIncrementally(
+		nodeIdentifier: TreeNodeSchemaIdentifier,
+		fieldKey: FieldKey,
+	): boolean;
+	/**
+	 * Called to encode an incremental field at the cursor. The chunks for this field are encoded separately
+	 * from the main buffer.
+	 * @param cursor - The cursor pointing to the field to encode.
+	 * @param chunkEncoder - A function that encodes the contents of the passed chunk in the field.
+	 * @returns The reference IDs of the encoded chunks in the field. This is used to retrieve the encoded
+	 * chunks later.
+	 */
+	encodeIncrementalField(
+		cursor: ITreeCursorSynchronous,
+		chunkEncoder: (chunk: TreeChunk) => EncodedFieldBatch,
+	): ChunkReferenceId[];
+}
+
+/**
+ * Properties for incremental decoding.
+ * Fields that had their chunks incrementally encoded will retrieve these chunks by calling
+ * `getEncodedIncrementalChunk`. See {@link IncrementalEncoder} for more details.
+ * @internal
```

Comment:
this does not appear to be package exported, so it should not be marked as internal: the internal tag is for package exported APIs which can only be used by other packages in our repo.

I think this applies to some other APIs in this PR as well.